compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -1373,8 +1375,20 @@ private boolean validateErrorVariable(BLangErrorVariable errorVariable, BErrorTy
         }
 
         if (isRestDetailBindingAvailable(errorVariable)) {
+            // Type of rest pattern is a map type where constraint type is,
+            // union of keys whose values are not matched in error binding/match pattern.
             BTypeSymbol typeSymbol = createTypeSymbol(SymTag.TYPE);
-            BMapType restType = new BMapType(TypeTags.MAP, recordType.restFieldType, typeSymbol);
+            BUnionType restFieldType = BUnionType.create(null, recordType.restFieldType);
+            for (Map.Entry<String, BField> entry : fieldMap.entrySet()) {
+                if (!matchedDetailFields.contains(entry.getKey())) {
+                    BType type = entry.getValue().getType();
+                    if (!type.equals(symTable.nilType)) {
+                        restFieldType.add(type);
```

Comment:
Shouldn't we check if it is assignable to the current collection of types and add only if not? 

@gimantha is doing something similar in https://github.com/ballerina-platform/ballerina-lang/pull/20180/files#diff-782094a79e6831ac57a34ad157af49c1R1177 to identify the narrowest type representing all possible fields.

Reply:
Fixed